<?xml version='1.0'?>
<robot name="artpark_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:include filename="$(find artpark_description)/urdf/artpark_robot.gazebo" />
  <xacro:include filename="$(find artpark_description)/urdf/materials.xacro" />

  <gazebo>
    <static>false</static>
  </gazebo>

  <link name="base_link"/>

  <link name="chasis">
    <collision name='chasis_collision'>
      <geometry>
        <box size=".6 .6 .1"/>
      </geometry>
    </collision>
    <visual name='chasis_visual'>
      <geometry>
        <box size=".6 .6 .1"/>
      </geometry>
      <material name="red" />
    </visual>
    <inertial name="chasis_link_inertial">
      <mass value="2"/>
      <inertia ixx="0.06167" ixy="0.0" ixz="0.0" iyy="0.06167" iyz="0.0" izz="0.12"/>
    </inertial>
  </link>

  <joint name="chasis_joint" type="fixed">
    <parent link="base_link"/>
    <child link="chasis"/>
    <origin xyz="0 0 0.15" rpy="0 0 0"/>
  </joint>
  
  <link name="body">
    <collision name='body_collision'>
      <geometry>
        <box size=".6 .6 .8"/>
      </geometry>
    </collision>
    <visual name='body_visual'>
      <geometry>
        <box size=".6 .6 .8"/>
      </geometry>
      <material name="red" />
    </visual>
    <inertial name="base_link_inertial">
      <mass value="2"/>
      <inertia ixx="0.16667" ixy="0.0" ixz="0.0" iyy="0.16667" iyz="0.0" izz="0.12"/>
    </inertial>
  </link>

  <joint name="body_joint" type="continuous">
    <parent link="chasis"/>
    <child link="body"/>
    <origin xyz="0 0 .45" rpy="0 0 0"/>
    <axis xyz="0.0 0.0 1.0"/>
    <dynamics damping="1"/>
  </joint>

  <transmission name="body_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="body_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="body_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="camera">
    <collision name="camera_collision">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual name="camera_visual">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="white" />
    </visual>
    <inertial name="camera_inertial">
      <mass value="0.1"/>
      <inertia ixx="0.00000167" ixy="0.0" ixz="0.0" iyy="0.00000167" iyz="0.0" izz="0.00000167"/>
    </inertial>
  </link>

  <joint name="camera_joint" type="fixed">
    <parent link="body"/>
    <child link="camera"/>
    <origin xyz=".3 0 .015" rpy="0.0 0.0 0.0"/>   <!--rpy = "0 .4 0"-->
  </joint>

  <link name="sonar_front">
    <collision name="front_sonar_collision">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual name="front_sonar_visual">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="white" />
    </visual>
    <inertial name="front_sonar_inertial">
      <mass value="0.01"/>
      <inertia ixx="0.000000167" ixy="0.0" ixz="0.0" iyy="0.000000167" iyz="0.0" izz="0.000000167"/>
    </inertial>
  </link>

  <joint name="front_sonar_joint" type="fixed">
    <parent link="body"/>
    <child link="sonar_front"/>
    <origin xyz=".3 0 .395" rpy="0.0 0.0 0.0"/>
  </joint>

  <link name="sonar_front_left">
    <collision name="front_left_sonar_collision">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual name="front_left_sonar_visual">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="white" />
    </visual>
    <inertial name="front_left_sonar_inertial">
      <mass value="0.01"/>
      <inertia ixx="0.000000167" ixy="0.0" ixz="0.0" iyy="0.000000167" iyz="0.0" izz="0.000000167"/>
    </inertial>
  </link>

  <joint name="front_left_sonar_joint" type="fixed">
    <parent link="body"/>
    <child link="sonar_front_left"/>
    <origin xyz=".3 0.295 .395" rpy="0.0 0.0 0.7854"/>
  </joint>

  <link name="sonar_front_right">
    <collision name="front_right_sonar_collision">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual name="front_right_sonar_visual">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="white" />
    </visual>
    <inertial name="front_right_sonar_inertial">
      <mass value="0.01"/>
      <inertia ixx="0.000000167" ixy="0.0" ixz="0.0" iyy="0.000000167" iyz="0.0" izz="0.000000167"/>
    </inertial>
  </link>

  <joint name="front_right_sonar_joint" type="fixed">
    <parent link="body"/>
    <child link="sonar_front_right"/>
    <origin xyz=".3 -0.295 .395" rpy="0.0 0.0 -0.7854"/>
  </joint>

  <link name="sonar_left">
    <collision name="left_sonar_collision">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual name="left_sonar_visual">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="white" />
    </visual>
    <inertial name="left_sonar_inertial">
      <mass value="0.01"/>
      <inertia ixx="0.000000167" ixy="0.0" ixz="0.0" iyy="0.000000167" iyz="0.0" izz="0.000000167"/>
    </inertial>
  </link>

  <joint name="left_sonar_joint" type="fixed">
    <parent link="body"/>
    <child link="sonar_left"/>
    <origin xyz="0.0 0.3 .395" rpy="0.0 0.0 1.57"/>
  </joint>

  <link name="sonar_right">
    <collision name="right_sonar_collision">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual name="right_sonar_visual">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="white" />
    </visual>
    <inertial name="right_sonar_inertial">
      <mass value="0.01"/>
      <inertia ixx="0.000000167" ixy="0.0" ixz="0.0" iyy="0.000000167" iyz="0.0" izz="0.000000167"/>
    </inertial>
  </link>

  <joint name="right_sonar_joint" type="fixed">
    <parent link="body"/>
    <child link="sonar_right"/>
    <origin xyz="0.0 -0.3 .395" rpy="0.0 0.0 -1.57"/>
  </joint>

  <link name="sonar_rear">
    <collision name="rear_sonar_collision">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual name="rear_sonar_visual">
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="white" />
    </visual>
    <inertial name="rear_sonar_inertial">
      <mass value="0.01"/>
      <inertia ixx="0.000000167" ixy="0.0" ixz="0.0" iyy="0.000000167" iyz="0.0" izz="0.000000167"/>
    </inertial>
  </link>

  <joint name="rear_sonar_joint" type="fixed">
    <parent link="body"/>
    <child link="sonar_rear"/>
    <origin xyz="-0.3 0.0 -0.35" rpy="0.0 0.0 3.14"/>
  </joint>

  <link name="mop">
    <collision name='mop_collision'>
      <geometry>
        <cylinder radius="0.25" length="0.1"/>
      </geometry>
    </collision>
    <visual name='mop_visual'>
      <geometry>
        <mesh filename="file://$(find artpark_description)/meshes/mop.dae" scale="1.25 1.25 1"/>
      </geometry>
      <material name="green" />
    </visual>
    <inertial name="mop_inertial">
      <mass value="0.15"/>
      <inertia ixx="0.00246875" ixy="0.0" ixz="0.0" iyy="0.00246875" iyz="0.0" izz="0.0046875"/>
    </inertial>
  </link>

  <joint name="mop_joint" type="fixed">
    <parent link="base_link"/>
    <child link="mop"/>
    <origin xyz="0 0 0.055" rpy="0 0 0"/>
  </joint>

  <link name="trash_holder">
    <collision name='trash_holder_collision'>
      <origin xyz="0.15 0 0.1" rpy="0.0 0.0 1.57"/>
      <geometry >
        <mesh filename="file://$(find artpark_description)/meshes/trash_holder.dae"/>
      </geometry>
    </collision>
    <visual name='trash_holder_visual'>
      <origin xyz="0.15 0 0.1" rpy="0.0 0.0 1.57"/>
      <geometry>
        <mesh filename="file://$(find artpark_description)/meshes/trash_holder.dae"/>
      </geometry>
      <material name="black" />
    </visual>
    <inertial name="trash_holder_inertial">
      <origin xyz="0.15 0 0.1" rpy="0.0 0.0 1.57"/>
      <mass value="0.1"/>
      <inertia ixx="0.001083" ixy="0.0" ixz="0.0" iyy="0.00333" iyz="0.0" izz="0.00375"/>
    </inertial>
  </link>

  <joint name="trash_holder_joint" type="revolute">
    <parent link="body"/>
    <child link="trash_holder"/>
    <origin xyz="-0.3 0 0.4" rpy="0.0 0.0 0"/>
    <axis xyz="0.0 1.0 0.0"/>
    <limit effort="0.1" lower="-2.7475" upper="0" velocity="0.1"/>
  </joint>

  <transmission name="trash_holder_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="trash_holder_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="trash_holder_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_base">
    <collision name="arm_base_collision">
      <geometry>
        <cylinder radius="0.1" length="0.2"/>
      </geometry>
    </collision>
    <visual name="arm_base_visual">
      <geometry>
        <mesh filename="file://$(find artpark_description)/meshes/arm_base.dae"/>
      </geometry>
    </visual>
    <inertial name="arm_base_inertial">
      <mass value="0.6"/>
      <inertia ixx="0.0035" ixy="0.0" ixz="0.0" iyy="0.0035" iyz="0.0" izz="0.003"/>
    </inertial>
  </link>

  <joint name="arm_base_joint" type="revolute">
    <parent link="body"/>
    <child link="arm_base"/>
    <origin xyz="0.15 0.0 0.4" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 0.0 1.0"/>
    <limit effort="0.1" lower="-1.57" upper="1.57" velocity="1.0"/>
    <dynamics damping="1.0"/>
  </joint>

  <transmission name="arm_base_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="arm_base_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="arm_base_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="front_left_caster">
    <collision name='front_left_caster_collision'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <visual name='front_left_caster_visual'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </visual>
    <inertial name="front_left_caster_inertial">
      <mass value="0.1"/>
      <inertia ixx="0.000167" ixy="0.0" ixz="0.0" iyy="0.000167" iyz="0.0" izz="0.000167"/>
    </inertial>
  </link>

  <joint name="front_left_caster_joint" type="continuous">
    <parent link="chasis"/>
    <child link="front_left_caster"/>
    <origin xyz=".25 .25 0" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 0.0 1.0"/>
    <dynamics damping="1.0" />
  </joint>

  <transmission name="front_left_caster_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_left_caster_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="front_left_caster_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="front_left_wheel">
    <collision name='front_left_wheel_collision'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </collision>
    <visual name='front_left_wheel_visual'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </visual>
    <inertial name="front_left_wheel_inertial">
      <mass value="1"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="front_left_wheel_joint" type="continuous">
    <parent link="front_left_caster"/>
    <child link="front_left_wheel"/>
    <origin xyz="0 0 -0.1" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 1.0 0.0"/>
  </joint>

  <transmission name="front_left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_left_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="front_left_wheel_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="front_right_caster">
    <collision name='front_right_caster_collision'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <visual name='front_right_caster_visual'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </visual>
    <inertial name="front_right_caster_inertial">
      <mass value="0.1"/>
      <inertia ixx="0.000167" ixy="0.0" ixz="0.0" iyy="0.000167" iyz="0.0" izz="0.000167"/>
    </inertial>
  </link>

  <joint name="front_right_caster_joint" type="continuous">
    <parent link="chasis"/>
    <child link="front_right_caster"/>
    <origin xyz=".25 -.25 0" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 0.0 1.0"/>
    <dynamics damping="1.0" />
  </joint>

  <transmission name="front_right_caster_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_right_caster_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="front_right_caster_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="front_right_wheel">
    <collision name='front_right_wheel_collision'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </collision>
    <visual name='front_right_wheel_visual'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </visual>
    <inertial name="front_right_wheel_inertial">
      <mass value="1"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="front_right_wheel_joint" type="continuous">
    <parent link="front_right_caster"/>
    <child link="front_right_wheel"/>
    <origin xyz="0 0 -0.1" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 1.0 0.0"/>
  </joint>

  <transmission name="front_right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_right_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="front_right_wheel_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="rear_left_caster">
    <collision name='rear_left_caster_collision'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <visual name='rear_left_caster_visual'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </visual>
    <inertial name="rear_left_caster_inertial">
      <mass value="0.1"/>
      <inertia ixx="0.000167" ixy="0.0" ixz="0.0" iyy="0.000167" iyz="0.0" izz="0.000167"/>
    </inertial>
  </link>

  <joint name="rear_left_caster_joint" type="continuous">
    <parent link="chasis"/>
    <child link="rear_left_caster"/>
    <origin xyz="-.25 .25 0" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 0.0 1.0"/>
    <dynamics damping="1.0" />
  </joint>

  <transmission name="rear_left_caster_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rear_left_caster_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rear_left_caster_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="rear_left_wheel">
    <collision name='rear_left_wheel_collision'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </collision>
    <visual name='rear_left_wheel_visual'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </visual>
    <inertial name="rear_left_wheel_inertial">
      <mass value="1"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="rear_left_wheel_joint" type="continuous">
    <parent link="rear_left_caster"/>
    <child link="rear_left_wheel"/>
    <origin xyz="0 0 -0.1" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 1.0 0.0"/>
  </joint>

  <transmission name="rear_left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rear_left_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rear_left_wheel_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="rear_right_caster">
    <collision name='rear_right_caster_collision'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <visual name='rear_right_caster_visual'>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </visual>
    <inertial name="rear_right_caster_inertial">
      <mass value="0.1"/>
      <inertia ixx="0.000167" ixy="0.0" ixz="0.0" iyy="0.000167" iyz="0.0" izz="0.000167"/>
    </inertial>
  </link>

  <joint name="rear_right_caster_joint" type="continuous">
    <parent link="chasis"/>
    <child link="rear_right_caster"/>
    <origin xyz="-.25 -.25 0" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 0.0 1.0"/>
    <dynamics damping="1.0" />
  </joint>

  <transmission name="rear_right_caster_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rear_right_caster_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rear_right_caster_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="rear_right_wheel">
    <collision name='rear_right_wheel_collision'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </collision>
    <visual name='rear_rigth_wheel_visual'>
      <geometry>
        <sphere radius=".05"/>
      </geometry>
    </visual>
    <inertial name="rear_right_wheel_inertial">
      <mass value="1"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="rear_right_wheel_joint" type="continuous">
    <parent link="rear_right_caster"/>
    <child link="rear_right_wheel"/>
    <origin xyz="0 0 -0.1" rpy="0.0 0.0 0.0"/>
    <axis xyz="0.0 1.0 0.0"/>
  </joint>

  <transmission name="rear_right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rear_right_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rear_right_wheel_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

</robot>